### PR TITLE
feat: Support LISTEN broadcast in Realtime

### PR DESCRIPTION
**Pull request by filipecabaco**
## What kind of change does this PR introduce?

Support LISTEN/NOTIFY on `realtime:broadcast` which allows users to trigger broadcast messages by sending a payload to this topic.

```sql
select pg_notify(
          'realtime:broadcast',
          json_build_object(
              'private', $1::boolean,
              'topic', $2::text,
              'event', $3::text,
              'payload', $4::jsonb
          )::text
      );
```

We want to keep it as close to the Broadcast endpoint as possible and it already includes the feature flag `notify_private_alpha` to enabled it (defaults to false)

### CODE REVIEW

File: lib/realtime/tenants/connect.ex [hunk 1]
```diff
@@ -121,8 +124,26 @@ defmodule Realtime.Tenants.Connect do
...
+  end
 
-    {:noreply, state}
+  def handle_continue(
+        :setup_listen,
+        %{tenant_id: tenant_id} = state
+      ) do
+    tenant = Tenants.get_tenant_by_external_id(tenant_id)
+
+    if tenant.notify_private_alpha do
```
Comment (chasers) [inline lib/realtime/tenants/connect.ex:136]: Nice

File: lib/realtime/tenants/connect.ex [hunk 2]
```diff
@@ -164,6 +185,26 @@ defmodule Realtime.Tenants.Connect do
     {:noreply, state}
   end
 
+  def handle_info(
+        {:notification, _, _, "realtime:broadcast", payload},
```
Comment (chasers) [inline lib/realtime/tenants/connect.ex:189]: Feels like this is odd in the connect module. Too much responsibility for it, but nit for now.

File: lib/realtime_web/controllers/broadcast_controller.ex [hunk 3]
```diff
@@ -43,89 +32,8 @@ defmodule RealtimeWeb.BroadcastController do
...
-            Enum.each(events, fn %{topic: sub_topic, payload: payload, event: event} ->
-              send_message_and_count(tenant, sub_topic, event, payload, false)
-            end)
-
-          _ ->
-            nil
-        end
-      end)
-
+    with :ok <- Realtime.Tenants.BatchBroadcast.broadcast(conn, tenant, attrs) do
```
Comment (chasers) [inline lib/realtime_web/controllers/broadcast_controller.ex:35]: Good helper

**Comment by abc3** [comment]
@filipecabaco when does the `Listen` connection stop?

**Comment by filipecabaco** [comment]
> @filipecabaco when does the `Listen` connection stop?

The processes are linked so I would expect it to stop the connection but worth to check! I will try to do a test to check the behaviour for sanity sake 👍

File: lib/realtime/tenants/connect.ex (hunk 2 above)
Comment (filipecabaco) [reply lib/realtime/tenants/connect.ex:189]: agreed, will try to check what I can do to move this out of connect (start the db conn here and move the listen somewhere else 🤔 )

**Review by chasers — CHANGES_REQUESTED** [review]
I was thinking this should be a separate extension where `event` would be `listen_notify` or similar.

And users would have to include this `extension` in their RLS policies. 

I feel like this warrants a separate section in the docs and this would make it a real "extension".

Clients should not be able to send each other `listen_notify` messages so we could filter these out this way also.

**Review by chasers — APPROVED** [review]

**Comment by filipecabaco** [comment]
@abc3 good call on the test, without restart: :transient we had the process auto restart wrongly, added a test to check what happens when we suspend a tenant (which kills conns)

**Comment by psteinroe** [comment]
just saw this on my timeline, and I think it should be noted in the docs for this that there is a relatively low limit on the notify payload size.

**Comment by filipecabaco** [comment]
@psteinroe yes, 8kb. we want to give a new channel of communication that is more native to the database vs pg_net

we'll need to be careful to properly explain in the docs and this won't be launched for a while as this will be "private alpha" and feature flagged

**Pull request merged**